pennylane/qchem/tapering.py

```diff
@@ -538,3 +539,142 @@ def taper_hf(generators, paulixops, paulix_sector, num_electrons, num_wires):
+    ...                             generators, paulixops, paulix_sector, H.wires)
+    ...     return qml.expval(qml.PauliZ(0)@qml.PauliZ(1))
+    >>> drawer = qml.draw(circuit, show_all_wires=True)
+    >>> print(drawer(params=[0.38686753]))
+        0: ─╭RXY(-0.10+0.00j)─╭RYX(-0.10+0.00j)─┤ ╭<Z@Z>
+        1: ─╰RXY(-0.10+0.00j)─╰RYX(-0.10+0.00j)─┤ ╰<Z@Z>
+
+    .. details::
+        :title: Theory
+
+        Consider :math:`G` to be the generator of a unitrary :math:`V(\theta)`, i.e.,
+
+        .. math::
+
+            V(\theta) = e^{i G \theta}.
+
+        Then, for :math:`V` to have a non-trivial and compatible tapering with the generators of symmetry
+        :math:`tau`, we should have :math:`[V, \tau_i] = 0\ \forall \theta \in \mathbb{R},\ \forall \tau_i`.
+        This would hold only when its generator itself commutes with each :math:`tau_i`,
+
+        .. math::
+
+            [V, \tau_i] = 0 \iff [G, \tau_i]\quad \forall \theta \in \mathbb{R},\ \forall \tau_i.
+
+        By ensuring this, we can taper the generator :math:`G` using the Clifford operators :math:`U`,
+        and exponentiate the transformed generator :math:`G^{\prime}` to obtain a tapered unitary
+        :math:`V^{\prime}`,
+
+        .. math::
+
+            V^{\prime} \equiv e^{i U^{\dagger} G U \theta} = e^{i G^{\prime} \theta}.
+
+        This directly extends to non-parameterized gates, as these gates can be written as a parametrized
+        gate with some fixed parameter :math:`\phi`. For example, applying :func:`~.PauliX` gate is equivalent
+        to applying :func:`~.RX` gate with :math:`\phi=\pi`.
+    """
+
+    if gen_op is None:
+        if operation.num_params < 1:  # Non-parameterized gates
+            gen_mat = -1j * scipy.linalg.logm(qml.matrix(operation, wire_order=wire_order))
+            gen_op = qml.Hamiltonian(
+                *qml.utils.decompose_hamiltonian(gen_mat, wire_order=wire_order, hide_identity=True)
+            )
+            qml.simplify(gen_op)
+            if gen_op.ops[0].label() == qml.Identity(wires=[wire_order[0]]).label():
+                gen_op -= qml.Hamiltonian([gen_op.coeffs[0]], [qml.Identity(wires=wire_order[0])])
+        else:  # Single-parameter gates
+            try:
+                gen_op = qml.generator(operation, "hamiltonian")
+            except ValueError as exc:
+                raise NotImplementedError(
+                    f"Generator for {operation} is not implemented, please provide it with 'gen_op' args."
+                ) from exc
+    else:
+        if not isinstance(gen_op, qml.Hamiltonian):
+            raise ValueError(
+                f"Generator for the operation needs to a qml.Hamiltonian, but got {type(gen_op)}."
+            )
+        mat1 = scipy.linalg.expm(1j * qml.matrix(gen_op, wire_order=wire_order))
+        mat2 = qml.matrix(operation, wire_order=wire_order)
+        phase = np.divide(mat1, mat2, out=np.zeros_like(mat1, dtype=complex), where=mat1 != 0)[
```

Comment:
I wonder if there is a smarter way to obtain the phase, and whether it matters. If the generator matrix is obtained from `qml.generator()`, then shouldn't `expm(1j * matrix)` give back the right matrix for the unitary describing the operation?

Reply:
Oh I see, this is in case the user provider their own? Maybe consider adding a short comment to the code, e.g.,
`# check that user-provided generator is correct`